### PR TITLE
feat(protocol-designer): allow FFs to be set via query params

### LINKED ISSUE #7483: PD: allow feature flags to be set by URL params
# Overview

Problem: when we do user tests, we have to always make a special "DO NOT MERGE" branch where the relevant project's feature flag is hard-coded to "on"

Other problem: in Cypress tests, E2E test relying on a FF to be on can't pass CI. One approach is #7439 -- this ticket is an alternative

# Proposal

Use URL params to override initial FF settings, including prerelease mode?? Something like `localhost:8080?OT_PD_PRERELEASE_MODE=1&OT_PD_OTHER_THING=1`

Then you can do `cy.visit('/?OT_PD_PRERELEASE_MODE=1&OT_PD_OTHER_THING=1')` -- I think?

### PR DESCRIPTION
**Pull request by shlokamin**
# Overview
This PR enables users to set FF values via query params. Ex: `designer.opentrons.com/?PRERELEASE_MODE=1`

Closes #7483

# Changelog
- Allow FFs to be set via query params

# Review requests

- Open up the branch incognito (all ffs should be set to false)
- Set a ff (like PRERELEASE_MODE) to true: via query params `designer.opentrons.com/?PRERELEASE_MODE=1`
- You should see the FF enabled

~~Bcuz FFs get stored and rehydrated, it was not good enough to just add the FFs passed in via query params into the`featureFlags` reducer on initialization, since they would just get overwritten. If y'all know of a cleaner way to do this(instead of from inside `configureStore`) lmk!~~

Update, see comments below. Now we just override FFs coming from selectors (and in the case of step generation, `localStorage`)

# Risk assessment
Low

### CODE REVIEW

File: protocol-designer/src/feature-flags/reducers.js [hunk 1]
```diff
@@ -19,10 +41,14 @@ import type { Action } from '../types'
 // initial values. Eg `OT_PD_PRERELEASE_MODE=1 make -C protocol-designer dev`
 // will initialize PRERELEASE_MODE to true (but as per the note above, that
 // initial value is only relevant if there is no persisted value already)
+
+/* $FlowFixMe(sa, 2021-03-26): Flow thinks getFlagsFromQueryParams will overwrite explicit properties 
+/* on initialFlags even though the return type of getFlagsFromQueryParams is type Flags */
```
Comment (shlokamin) [inline protocol-designer/src/feature-flags/reducers.js]: :(

**Review by Kadee80 — APPROVED** [review]
🍨

**Review by IanLondon — COMMENTED** [review]
Awesome!

This just has one weird aspect IMO -- if you go to a URL that uses the query params, that feature flag setting is persisted in your browser and will follow you around if you go to other branches in the domain. So eg if we run a user test on sandbox `user-research-A/?SOMETHING=1` and later the same user is directed to a follow-up `user-research-B/`, then the SOMETHING flag will still be left on even though we didn't necessarily intend it (unless user went in the UI and toggled it off while in sandbox subdomain). I wonder if we could make it so maybe a selector (or a `useFeatureFlags` hook?) reads the URL params, instead of dispatching an action to set them? Then the URL params would function as "overrides" but visiting a URL with params wouldn't modify the persisted localStorage FF state and act totally transiently.

Also a not-very-important weirdness: IDK why but the FF `PRERELEASE_MODE` doesn't follow the convention for other FFs where they start with `OT_PD_*` (it's been that way a while, not new in this PR!). I think that's okay though, we'll just have to remember that when using the URL param it's `?PRERELEASE_MODE=1` yet while using the env var it's eg `make -C protocol-designer dev PD_OT_PRERELEASE_MODE=1`. I think that's fine, we'll probably use the env vars less and less now that we have the URL params.

**Comment by shlokamin** [comment]
That is a good point, I like the the idea of query params overriding, but not persisting.

**Review by IanLondon — APPROVED** [review]
weird, maybe I was looking at an old version somehow! The other day when I tried it, I saw in the inspector that localStorage was being written to but now it isn't persisting so all good

**Pull request merged**